### PR TITLE
Send notifications of new posts in a community to slack

### PR DESCRIPTION
**Pull request by Connoropolous**
Woohoo! This was fun :)

I'm sure you'll have a few little pointers about things that need to be resolved before it gets merged, let me know what they are!

relates to front end PR https://github.com/Hylozoic/hylo-frontend/pull/85 
and issue #50 

- [x] point to a better image asset

screen shot 2015-11-19 at 5 23 12 pm

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -78,6 +78,7 @@
     "randomstring": "^1.1.0",
     "rc": "~0.5.0",
     "request": "^2.51.0",
+    "request-promise": "^1.0.2",
```
Comment (levity) [inline package.json]: interesting -- i've just been using bluebird's `promisify` on request.get and request.post, so i'm not sure if we need to add this dependency

Comment (Connoropolous) [inline package.json]: Ok cook. Yeah, that should work. Can you point me to example? I'm pretty new to promises fyi

File: migrations/20151119113256_community_slackhook.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+'use strict';
+
+exports.up = function(knex, Promise) {
+  return knex.schema.table('community', function(table) {
+    table.text('slack_hook');
```
Comment (levity) [inline migrations/20151119113256_community_slackhook.js]: interesting note -- i was going to suggest that `string` is more appropriate here than `text`, but then i looked it up and in Postgres [there's actually no difference](http://www.postgresql.org/docs/8.4/static/datatype-character.html). so carry on :)

**Comment by Connoropolous** [comment]
We should point to a better image asset than the one I pointed to.

File: api/models/Post.js [hunk 3]
```diff
@@ -236,18 +236,24 @@ module.exports = bookshelf.Model.extend({
...
-          alertText = PushNotification.textForNewPost(post, community, user.get('id'))
-          return user.sendPushNotification(alertText, path)
-        })
+        return Promise.join(
+          Promise.map(communities.models, (community) => {
+            if (!community.get('slack_hook')) return
+            return Community.sendSlackNotification(community.id, post); 
+          }),
+          Promise.map(users, (user) => {
+	    if (!user.get('push_new_post_preference')) return
```
Comment (levity) [inline api/models/Post.js]: your indentation seems to have gotten a little wonky here

File: package.json (hunk 1 above)
Comment (levity) [inline package.json]: Basically you just do:

```js
var Promise = require('bluebird')
var request = require('request')
var get = Promise.promisify(request.get)

get(url).spread((resp, body) => ...).catch(err => ...)
```

FYI: in bluebird 3 (which I started using in a different project; here we're still on v2) you have to do `promisify(get, {multiArgs: true})` in order to get both `resp` and `body` into the `then` handler

File: api/models/Post.js [hunk 4]
```diff
@@ -236,18 +236,24 @@ module.exports = bookshelf.Model.extend({
...
-          var postCommunitiesIds = communities.models.map(community => community.get('id'))
-          var community, path, alertText
-          community = _.find(userCommunities, community => _.contains(postCommunitiesIds, community.get('id')))
-          if (!community) return
-          path = url.parse(Frontend.Route.post(post, community)).path
-          alertText = PushNotification.textForNewPost(post, community, user.get('id'))
-          return user.sendPushNotification(alertText, path)
-        })
+        return Promise.join(
+          Promise.map(communities.models, (community) => {
```
Comment (levity) [inline api/models/Post.js]: just a little style note: you don't need to have `( )` for single-argument methods

**Comment by levity** [comment]
re: images: you could try these two

http://cdn.hylo.com/misc/hylo-logo-teal-on-transparent.png
http://cdn.hylo.com/misc/hylo-logo-white-on-teal-circle.png

**Comment by Connoropolous** [comment]
@levity I'm going to come back and finish this soon! oops! took me a bit. 

@edwardwest , the newly created CTA slack + CTA hylo is another example of a community who could benefit from this integration

**Comment by Connoropolous** [comment]
I am going to try to catch this up to master and then finish it :) finally. might ping for help if I get stuck doing migrations or something

**Comment by levity** [comment]
:cool:

**Comment by Connoropolous** [comment]
using the default image, it looks like this now

screen shot 2016-02-10 at 11 39 19 am

**Comment by Connoropolous** [comment]
This is ready for another review now :) 

The only steps needed to make this work would be 
- [ ] running the migration
- [ ] setting the HYLO_SLACK_ICON env variable

File: api/services/Slack.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
+var format = require('util').format
+var Promise = require('bluebird')
+var request = require('request')
+var post = Promise.promisify(request.post)
```
Comment (harlantwood) [inline api/services/Slack.js]: You may like to consider using the `request-promisify` module, which combines request + bluebird.

Comment (Connoropolous) [inline api/services/Slack.js]: haha, actually I had ben doing it that way, but Lawrence mentioned he'd been doing it this way in other places in the app, so I figured I'd just keep on that way :) cheers though Harlan. Keen for other pointers whenever you have them

Comment (harlantwood) [inline api/services/Slack.js]: Ah cool.  Consistency is worth a lot : )

**Pull request closed without merging**